diff --git a/netlink/nss_nlipv4.c b/netlink/nss_nlipv4.c
index 9778cdb..5951c77 100644
--- a/netlink/nss_nlipv4.c
+++ b/netlink/nss_nlipv4.c
@@ -69,6 +69,9 @@ struct nss_nlipv4_ctx {
 static int nss_nlipv4_ops_create_rule(struct sk_buff *skb, struct genl_info *info);
 static int nss_nlipv4_ops_destroy_rule(struct sk_buff *skb, struct genl_info *info);
 static int nss_nlipv4_process_notify(struct notifier_block *nb, unsigned long val, void *data);
+static void nss_nlipv4_process_event(void *app_data, struct nss_ipv4_msg *nim);
+static void nss_nlipv4_process_conn_sync_many(void *app_data, struct nss_ipv4_msg *nim);
+static void nss_nlipv4_mcast_nim(struct nss_ipv4_msg *nim);
 
 /*
  * multicast group for sending message status & events
@@ -653,6 +657,42 @@ static int nss_nlipv4_verify_identifier(struct nss_ipv4_rule_create_msg *msg)
 	return 0;
 }
 
+static void nss_nlipv4_mcast_nim(struct nss_ipv4_msg *nim)
+{
+	struct nss_nlipv4_rule *nl_rule;
+	struct sk_buff *skb;
+
+	skb = nss_nl_new_msg(&nss_nlipv4_family, NSS_NLCMN_SUBSYS_IPV4);
+	if (!skb) {
+		nss_nl_error("unable to allocate NSS_NLIPV4 event\n");
+		return;
+	}
+
+	nl_rule = nss_nl_get_data(skb);
+	memset(nl_rule, 0, sizeof(*nl_rule));
+	memcpy(&nl_rule->nim, nim, sizeof(nl_rule->nim));
+	nss_nl_mcast_event(&nss_nlipv4_family, skb);
+}
+
+static void nss_nlipv4_process_event(void *app_data, struct nss_ipv4_msg *nim)
+{
+	if (nim->cm.response != NSS_CMN_RESPONSE_NOTIFY)
+		return;
+
+	if (nim->cm.type != NSS_IPV4_RX_CONN_STATS_SYNC_MSG)
+		return;
+
+	nss_nlipv4_mcast_nim(nim);
+}
+
+static void nss_nlipv4_process_conn_sync_many(void *app_data, struct nss_ipv4_msg *nim)
+{
+	if (nim->cm.type != NSS_IPV4_TX_CONN_STATS_SYNC_MANY_MSG)
+		return;
+
+	nss_nlipv4_mcast_nim(nim);
+}
+
 /*
  * nss_nlipv4_process_notify()
  * 	process notification messages from NSS
@@ -670,6 +707,7 @@ static int nss_nlipv4_process_notify(struct notifier_block *nb, unsigned long va
 	}
 
 	nl_rule = nss_nl_get_data(skb);
+	memset(nl_rule, 0, sizeof(*nl_rule));
 	nss_stats = (struct nss_ipv4_stats_notification *)data;
 	memcpy(&nl_rule->stats, nss_stats, sizeof(struct nss_ipv4_stats_notification));
 	nss_nl_mcast_event(&nss_nlipv4_family, skb);
@@ -995,11 +1035,27 @@ bool nss_nlipv4_init(void)
 		goto unreg_family;
 	}
 
+	ret = nss_ipv4_notify_register_listener(nss_nlipv4_process_event, NULL);
+	if (ret) {
+		nss_nl_info_always("Error: unable to register IPv4 notify listener\n");
+		goto unreg_stats;
+	}
+
+	ret = nss_ipv4_conn_sync_many_notify_register_listener(nss_nlipv4_process_conn_sync_many);
+	if (ret) {
+		nss_nl_info_always("Error: unable to register IPv4 conn_sync_many listener\n");
+		goto unreg_notify;
+	}
+
 	return true;
 
 	/*
 	 * undo all registeration
 	 */
+unreg_notify:
+	nss_ipv4_notify_unregister_listener(nss_nlipv4_process_event, NULL);
+unreg_stats:
+	nss_ipv4_stats_unregister_notifier(&nss_ipv4_stats_notifier_nb);
 unreg_family:
 	genl_unregister_family(&nss_nlipv4_family);
 
@@ -1016,6 +1072,12 @@ bool nss_nlipv4_exit(void)
 
 	nss_nl_info_always("Exit NSS netlink IPv4 handler\n");
 
+	/*
+	 * unregister conn sync callbacks for ipv4
+	 */
+	nss_ipv4_notify_unregister_listener(nss_nlipv4_process_event, NULL);
+	nss_ipv4_conn_sync_many_notify_unregister_listener(nss_nlipv4_process_conn_sync_many);
+
 	/*
 	 * unregister the ops family
 	 */
diff --git a/netlink/nss_nlipv6.c b/netlink/nss_nlipv6.c
index 3223ac0..6936b2d 100644
--- a/netlink/nss_nlipv6.c
+++ b/netlink/nss_nlipv6.c
@@ -77,6 +77,9 @@ struct nss_nlipv6_ctx {
 static int nss_nlipv6_ops_create_rule(struct sk_buff *skb, struct genl_info *info);
 static int nss_nlipv6_ops_destroy_rule(struct sk_buff *skb, struct genl_info *info);
 static int nss_nlipv6_process_notify(struct notifier_block *nb, unsigned long val, void *data);
+static void nss_nlipv6_process_event(void *app_data, struct nss_ipv6_msg *nim);
+static void nss_nlipv6_process_conn_sync_many(void *app_data, struct nss_ipv6_msg *nim);
+static void nss_nlipv6_mcast_nim(struct nss_ipv6_msg *nim);
 
 /*
  * multicast group for sending message status & events
@@ -650,6 +654,42 @@ static int nss_nlipv6_verify_identifier(struct nss_ipv6_rule_create_msg *msg)
 	return 0;
 }
 
+static void nss_nlipv6_mcast_nim(struct nss_ipv6_msg *nim)
+{
+	struct nss_nlipv6_rule *nl_rule;
+	struct sk_buff *skb;
+
+	skb = nss_nl_new_msg(&nss_nlipv6_family, NSS_NLCMN_SUBSYS_IPV6);
+	if (!skb) {
+		nss_nl_error("unable to allocate NSS_NLIPV6 event\n");
+		return;
+	}
+
+	nl_rule = nss_nl_get_data(skb);
+	memset(nl_rule, 0, sizeof(*nl_rule));
+	memcpy(&nl_rule->nim, nim, sizeof(nl_rule->nim));
+	nss_nl_mcast_event(&nss_nlipv6_family, skb);
+}
+
+static void nss_nlipv6_process_event(void *app_data, struct nss_ipv6_msg *nim)
+{
+	if (nim->cm.response != NSS_CMN_RESPONSE_NOTIFY)
+		return;
+
+	if (nim->cm.type != NSS_IPV6_RX_CONN_STATS_SYNC_MSG)
+		return;
+
+	nss_nlipv6_mcast_nim(nim);
+}
+
+static void nss_nlipv6_process_conn_sync_many(void *app_data, struct nss_ipv6_msg *nim)
+{
+	if (nim->cm.type != NSS_IPV6_TX_CONN_STATS_SYNC_MANY_MSG)
+		return;
+
+	nss_nlipv6_mcast_nim(nim);
+}
+
 /*
  * nss_nlipv6_process_notify()
  *	process notification messages from NSS
@@ -667,6 +710,7 @@ static int nss_nlipv6_process_notify(struct notifier_block *nb, unsigned long va
 	}
 
 	nl_rule = nss_nl_get_data(skb);
+	memset(nl_rule, 0, sizeof(*nl_rule));
 	nss_stats = (struct nss_ipv6_stats_notification *)data;
 	memcpy(&nl_rule->stats, nss_stats, sizeof(struct nss_ipv6_stats_notification));
 	nss_nl_mcast_event(&nss_nlipv6_family, skb);
@@ -996,11 +1036,27 @@ bool nss_nlipv6_init(void)
 		goto unreg_family;
 	}
 
+	ret = nss_ipv6_notify_register_listener(nss_nlipv6_process_event, NULL);
+	if (ret) {
+		nss_nl_info_always("Error: unable to register IPv6 notify listener\n");
+		goto unreg_stats;
+	}
+
+	ret = nss_ipv6_conn_sync_many_notify_register_listener(nss_nlipv6_process_conn_sync_many);
+	if (ret) {
+		nss_nl_info_always("Error: unable to register IPv6 conn_sync_many listener\n");
+		goto unreg_notify;
+	}
+
 	return true;
 
 	/*
 	 * undo all registeration
 	 */
+unreg_notify:
+	nss_ipv6_notify_unregister_listener(nss_nlipv6_process_event, NULL);
+unreg_stats:
+	nss_ipv6_stats_unregister_notifier(&nss_ipv6_stats_notifier_nb);
 unreg_family:
 	genl_unregister_family(&nss_nlipv6_family);
 
@@ -1017,6 +1073,12 @@ bool nss_nlipv6_exit(void)
 
 	nss_nl_info_always("Exit NSS netlink IPV6 handler\n");
 
+	/*
+	 * Unregister conn sync callbacks for ipv6
+	 */
+	nss_ipv6_notify_unregister_listener(nss_nlipv6_process_event, NULL);
+	nss_ipv6_conn_sync_many_notify_unregister_listener(nss_nlipv6_process_conn_sync_many);
+
 	/*
 	 * Unregister the device callback handler for ipv6
 	 */
